### PR TITLE
Switch to PHP punycode implementation

### PR DESCRIPTION
**Pull request by mleko**
Switches from depending on ext-intl to true/punycode for the purposes of IDN to ASCII conversion. This change ensures that the functionality will work even when ext-intl is not available.

### CODE REVIEW

**Comment by glensc** [comment]
PR lacks basic information: "why?".

also, the intl dependency was optional, now you added replacement dependency as required.

you should also post what new dependencies got installed with this PR to help maintainers decide fate of this PR.

**Comment by mleko** [comment]
Composer dependency is easier to install than PECL extension. Also it don't require any special privileges.

**Comment by glensc** [comment]
don't "answer" it, put it to PR description!

File: src/Header/AbstractAddressList.php [hunk 1]
```diff
@@ -96,15 +102,19 @@ public function getFieldName()
...
     protected function idnToAscii($domainName)
     {
-        if (extension_loaded('intl')) {
-            return (idn_to_ascii($domainName) ?: $domainName);
+        if (null === self::$punycode) {
+            self::$punycode = new Punycode();
+        }
+        try {
+            return self::$punycode->encode($domainName);
+        } catch (\Exception $e) {
```
Comment (Ocramius) [inline src/Header/AbstractAddressList.php]: Is there no more specific exception type?

File: src/Header/AbstractAddressList.php [hunk 2]
```diff
@@ -96,15 +102,19 @@ public function getFieldName()
...
-     * @param string $domainName  the UTF-8 encoded email
+     * @param string $domainName the UTF-8 encoded email
      * @return string
      */
     protected function idnToAscii($domainName)
     {
-        if (extension_loaded('intl')) {
-            return (idn_to_ascii($domainName) ?: $domainName);
+        if (null === self::$punycode) {
+            self::$punycode = new Punycode();
```
Comment (Ocramius) [inline src/Header/AbstractAddressList.php]: `if (! self::$punycode) {`

File: src/Header/AbstractAddressList.php [hunk 3]
```diff
@@ -39,6 +40,11 @@
      */
     protected static $type;
 
+    /**
+     * @var Punycode
```
Comment (Ocramius) [inline src/Header/AbstractAddressList.php]: `|null`

Comment (mleko) [inline src/Header/AbstractAddressList.php]: Good point, thank you

File: src/Header/AbstractAddressList.php (hunk 1 above)
Comment (mleko) [inline src/Header/AbstractAddressList.php]: I'm not sure if it's beneficial but I found one

File: src/Header/AbstractAddressList.php (hunk 2 above)
Comment (mleko) [inline src/Header/AbstractAddressList.php]: I prefer my version. Is there style guide somewhere?

**Comment by weierophinney** [comment]
Thanks, @mleko!

**Pull request merged**